_posts/2016/2016-09-08-kubernetes-preparing-nodes.md

```diff
@@ -0,0 +1,259 @@
+
+## Flanneld
+
+[Flannel](https://coreos.com/flannel) enables the cluster wide distribution of the pods' packages.
+Since the description of the networks are stored in *etcd* a connection of all Kubernetes nodes (master and worker) to the *etcd* cluster has to be enabled.
+The docker daemon on each Kubernetes node will be performed with special parameters provided by flanneld.
+Cross-node communication of containers is realized by iptables that route the packages to the correct target servers.
+For this purpose flanneld creates its own virtual network interface and must therefore run with root rights.
+
+### Create *etcd* client certificates
+
+Kubernetes nodes need a client certificate because each user of the *etcd* cluster has to authenticate.
+It can be created like described in the first part of this blog post series but there are differences in the _openssl.cnf_.
+It is a bit simpler since flanneld has to show itself as client only.
+
+{% highlight bash %}
+root@kubernetes:~$ mkdir -p /etc/ssl/etcd
+root@kubernetes:~$ cd /etc/ssl/etcd
+root@kubernetes:/etc/ssl/etcd$ cat > openssl.cnf <<EOF
+[req]
+req_extensions = v3_req
+distinguished_name = req_distinguished_name
+[req_distinguished_name]
+[ v3_req ]
+basicConstraints = CA:FALSE
+keyUsage = nonRepudiation, digitalSignature, keyEncipherment
+extendedKeyUsage = clientAuth
+EOF
+root@kubernetes:/etc/ssl/etcd$ openssl genrsa -out etcd-client.key 2048
+root@kubernetes:/etc/ssl/etcd$ chmod 600 etcd-client.key
+root@kubernetes:/etc/ssl/etcd$ openssl \
+  req -new \
+  -key etcd-client.key \
+  -out etcd-client.csr \
+  -subj "/CN=$(hostname -s)" \
+  -extensions v3_req \
+  -config openssl.cnf \
+  -sha256
+root@kubernetes:/etc/ssl/etcd$ scp openssl.cnf etcd-client.csr root@root-ca-host:
+{% endhighlight %}
+
+This certificate request is also signed with the root CA stored as _/etc/ssl/etcd/etcd-client.crt_ on the Kubernetes nodes.
+
+{% highlight bash %}
+root@root-ca-host:~$ openssl x509 \
+  -req \
+  -CA ca.crt \
+  -CAkey ca.key \
+  -CAcreateserial \
+  -in etcd-client.csr \
+  -out etcd-client.crt \
+  -days 365 \
+  -extensions v3_req \
+  -extfile openssl.cnf \
+  -sha256
+root@root-ca-host:~$ scp etcd-client.crt ca.crt root@kubernetes:/etc/ssl/etcd/  
+{% endhighlight %}
+
+Alltogether there should be the following files on all Kubernetes nodes:
+
+{% highlight bash %}
+root@kubernetes:~$ cd /etc/ssl/etcd/
+root@kubernetes:/etc/ssl/etcd$ rm -f openssl.cnf etcd-client.csr
+root@kubernetes:/etc/ssl/etcd$ ls -la
+-rw-r--r--  1 root root 4711 Jun  33 08:15 ca.crt
+-rw-r--r--  1 root root 4712 Jun  33 08:15 etcd-client.crt
+-rw-------  1 root root 4713 Jun  33 08:15 etcd-client.key
+{% endhighlight %}
+
+### Installation of the daemon
+
+For the various settings an extra file __options.env__ is used.
+Besides all *Etcd* nodes it contains the certificates.
```

Comment:
*Etcd*  > *etcd* 

Besides all *etcd* nodes __options.env__ file contains the certificates.